hasjob/views/index.py

```diff
@@ -354,12 +301,93 @@ def index(basequery=None, md5sum=None, tag=None, domain=None, location=None, tit
                 # batch = [(pinned, post), ...]
                 loadmore = batch[-1][1].datetime
             pinsandposts = batch
-    if grouped:
+
+    query_params = request_args.to_dict(flat=False)
+    if loadmore:
+        query_params.update({'startdate': loadmore.isoformat() + 'Z', 'ph': pinned_hashids})
+    if location:
+        query_params.update({'l': location['name']})
+
+    if pay_graph:
+        pay_graph_data = make_pay_graph(pay_graph, posts, rmin=f_min, rmax=f_max)
+
+    return dict(posts=posts, pinsandposts=pinsandposts, grouped=grouped, newlimit=newlimit, title=title,
+        md5sum=md5sum, domain=domain, location=location, employer_name=employer_name,
+        showall=showall, f_locations=f_locations, loadmore=loadmore,
+        search_domains=search_domains, query_params=query_params,
+        data_filters=data_filters,
+        pay_graph_data=pay_graph_data, paginated=index_is_paginated(), template_vars=template_vars)
+
+
+@app.route('/', methods=['GET', 'POST'], subdomain='<subdomain>')
+@app.route('/', methods=['GET', 'POST'])
+@render_with({'text/html': 'index.html', 'application/json': json_index}, json=False)
+def index(basequery=None, md5sum=None, tag=None, domain=None, location=None, title=None, showall=True, statuses=None, batched=True, ageless=False, template_vars={}):
+    now = datetime.utcnow()
+    is_siteadmin = lastuser.has_permission('siteadmin')
+    board = g.board
+
+    if board:
+        board_jobs = {r.jobpost_id: r for r in
+            BoardJobPost.query.join(BoardJobPost.jobpost).filter(
+                BoardJobPost.board == g.board, JobPost.datetime > now - agelimit).options(
+                    db.load_only('jobpost_id', 'pinned')).all()
+        }
+
+    else:
+        board_jobs = {}
+
+    if basequery is None:
+        is_index = True
+    else:
+        is_index = False
+    if basequery is None and not (g.user or g.kiosk or (board and not board.require_login)):
+        showall = False
+        batched = False
+
+    search_query = request.args.get('q')
+    if search_query:
+        search_query = for_tsquery(request.args.get('q'))
+        try:
+            # TODO: Can we do syntax validation without a database roundtrip?
+            db.session.query(db.func.to_tsquery(search_query)).all()
```

Comment:
Let's push this into cache as well, so we don't have to validate by hitting the db every single time. This can be a separate cache region, `hasjob_query`, with validity period in days rather than hours. All it has is a list of query terms and validity status (so we cache both valid and invalid terms).